### PR TITLE
[#716] Fix possible NPE when no message is present

### PR DESCRIPTION
**Pull request by ctron**

### CODE REVIEW

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java [hunk 1]
```diff
@@ -39,6 +39,7 @@
  */
 public class DefaultFailureHandler implements Handler<RoutingContext> {
 
+    private static final String MSG_NO_ERROR = "<no error>";
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: I would rather send an empty string in this case. Getting a response with an error code that contains a body saying "no error" would be quite confusing, wouldn't it?

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java [hunk 2]
```diff
@@ -75,6 +76,32 @@ public void handle(final RoutingContext ctx) {
...
     }
 
+    private String extractMessage(final HttpServerResponse response) {
+        if ( response == null ) {
+            return MSG_NO_ERROR;
+        }
+
+        final String message = response.getStatusMessage();
+        if ( message == null ) {
+            return Integer.toString(response.getStatusCode());
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: this information is already given by the response's status code. I would rather use the empty string here again (i.e. an empty body)

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java [hunk 3]
```diff
@@ -75,6 +76,32 @@ public void handle(final RoutingContext ctx) {
...
+    }
+
+    private String extractMessage(final Throwable cause) {
+        if (cause == null) {
+            return MSG_NO_ERROR;
+        }
+
+        final String message = cause.getMessage();
+        if (message == null) {
+            return cause.getClass().getName();
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: I would rather not expose the exception class name here in order not to encourage people to actually rely on the class names...

Comment (ctron) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: I had this at first ... the problem with that seems to me that you would have no clue at all what went wrong. IMHO known the class name of an Exception doesn't really help you get any insight at all.

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java (hunk 2 above)
Comment (ctron) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: ok

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java (hunk 1 above)
Comment (ctron) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: Maybe "unknown error" is better. But having nothing at all again makes it hard to figure out what went wrong. Having some kind of string helps you find the right spot in the source code easily.

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java (hunk 3 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: Which view are you taking? A Hono developer? A user of a hosted Hono instance?

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java (hunk 1 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: How does this help?

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java [hunk 4]
```diff
@@ -39,6 +39,7 @@
  */
 public class DefaultFailureHandler implements Handler<RoutingContext> {
 
+    private static final String MSG_UNKNOWN_ERROR = "<unknown error>";
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: still, IMHO this is not helpful at all and I would rather have an empty body in this case

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java [hunk 5]
```diff
@@ -75,6 +76,32 @@ public void handle(final RoutingContext ctx) {
         }
     }
 
+    private String extractMessage(final HttpServerResponse response) {
+        if ( response == null ) {
+            return MSG_UNKNOWN_ERROR;
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: this code is unreachable, we will have run into an NPE in line 56 before we arrive here ...

File: service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java (hunk 3 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java]: A *user* shouldn't see the internal class names. A *developer* can take a look at the (server side) logs.

**Pull request closed without merging**